### ISSUE TITLE
Keener - duplicate quarterly fee b/c of two billings in July

### ISSUE THREAD

**Issue by kdjstudios** [kdjstudios/SABillingGitlab#1035]
In GitLab by @kdjstudios on Aug 13, 2018, 10:58

**Submitted by:** Gaylan Garrett <gaylan@keenercom.net>

**Helpdesk:** http://www.servicedesk.answernet.com/profiles/ticket/2018-08-09-82935/conversation

**Server:** External

**Client/Site:** Keener

**Account:** NA

**Issue:** 

Is there a way to prevent a duplicate billing of a quarterly fee that appears in a month that has two billing cycles ?

 

We have quarterly fees for the month of July and this year there was a billing done on 7/2 and then again on 7/30 and those clients received another invoice with another charge.  Example 6405.

 

Also with the staged fees intermittently processing twice, once when I create the invoice in between billing cycles and then again when the next billing is done, I am concerned that adding a staged fee to credit for the duplicate charge may result in yet another invoice being created with that credit again.

 

Any ideas / solutions ?

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Aug 14, 2018, 09:12

@smasih Can you pls detail the current mechanism that determines if a recurring fee should be assessed?  I would like to know how this works for monthly and other recurrence values.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Aug 14, 2018, 09:13

During discussion today, @tim.traylor mentioned we will need to have some sort of mechanism to prevent the duplicate transactions.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Aug 20, 2018, 02:02

@kdjstudios @tim.traylor 

* When Invoice creation is triggered 
* System checks if the account is active (because we cannot charge recurring fees on terminated account)
* Then system gathers all the child accounts and collect their recurring fees (including Partent's)
* This system check if the fee is disabled either on account or Site level.
* If the fee is active on both (account and site) level then processing moves further.
* In the next step system checks if the `fee is applicable`
 * This is checked by Current Billing Cycle's **service start date's month**
 * Each recurring fee has a field **Billing Period** 
 * ![image](/uploads/7ac917404f462441498ddcec80a075dd/image.png)
 * Billing Period are children to Billing Period on Site
 * ![image](/uploads/99fed6745ac210300ac04950061134eb/image.png)
 * If Billing Period on account is blank/NA then the `fee is applicable` every time the invoice is generated on account (manual or batch)
 * If **service start date's month** matches the label in Account's reccuring fees's Billing Period then the system will include the recurring charges each time the invoice is generated on the account irrespective of the generation process.

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Aug 20, 2018, 06:18

@smasih 
*  I propose we add a 'last_assessed_for_service_date' date to the recurring fees.
*  In the fee processing, at the step 'This is checked by Current Billing Cycle's **service start date's month**', we will also then check that the last_assessed_for_service_date date is NOT the current year and month (i.e., we did not already assess the fee this month).
*  If we did not already assess it we apply the fee and save the service start date in the last_assessed_for_service_date field.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Aug 21, 2018, 08:35

@smasih @pchaudhary 

Any updates on this?

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Aug 21, 2018, 10:24

@smasih Can we pls get an estimate on this one?

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Aug 23, 2018, 05:34

@tim.traylor 

Estimate `2.5hr`

Bifurcation of estimate:  
 * Setting up the system for development (like creating a branch)
 * development
 * developer level testing
 * testing with production data 
 * revisit the development if required 
 * buffer time  

approval required for this estimate.

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Aug 23, 2018, 10:24

@smasih This is approved to be fixed right away.  Thx

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Sep 7, 2018, 07:28

spent around 15-20min understanding the issue today.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Sep 11, 2018, 04:01

Hi [Tim] ( @tim.traylor ),   

As per the yesterday's huddle, we are still waiting for your response into this.

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Sep 13, 2018, 07:03

Hi [Kyle] (@kdjstudios ) & [Tim] (@tim.traylor ),   

Marking primary for #1098. We're still waiting for your response.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Sep 13, 2018, 07:03

assigned to @pchaudhary

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Sep 13, 2018, 08:22

@pchaudhary 

Marking for discussion, as I do not see what you are waiting on us for? @tim.traylor has already approved this.

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Sep 13, 2018, 09:46

Let's amend the previous design to his:

- I propose we add a 'last_assessed_for_service_date' date to the recurring fees.
- In the fee processing, at the step 'This is checked by Current Billing Cycle's **service start date's month**', we will also then check that the last_assessed_for_service_date date is NOT the current year and month (i.e., we did not already assess the fee this month).  
- If we did not already assess it we apply the fee and save the service start date in the last_assessed_for_service_date field.

*  If the fee was already assessed in the current year and month then we also need to check the account's "Turn on date" - if the turn on date year and month match the service period year and month then we will go ahead and assess the fee.  (This scenario addresses when an account first starts service and we do a pro-rated recurring fee for a partial month's service. In this situation the account would have had an out of cycle manual invoice created, and that manual invoice will have had the same service period as the invoice that is automatic for the new billing cycle.)

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Sep 17, 2018, 02:04

@kdjstudios @tim.traylor 

* Total estimate `2.5h`

* Time taken for internal discussion `0.5h`

* Development `estimate 1h`

* testing `estimate 1h`
 * **Test Cases**
  * Check if the recurring fee is repeating itself in same billing cycle by creating two manual invoices.
  * Check if the recurring fee is repeating itself in same billing cycle by creating one invoice from usage process and one manual.
  * Test quarterly recurring fee in Monthly billing cycle.
  * Test quarterly recurring fee in Quarterly billing cycle.
  * Test when account turn on date comes in action.

* **Process we will follow**

* ![WhatsApp_Image_2018-09-17_at_1.18.06_PM](/uploads/1a39d0865d42a3cf855f8fb237884241/WhatsApp_Image_2018-09-17_at_1.18.06_PM.jpeg)

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Sep 17, 2018, 02:04

changed time estimate to 2h 30m

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Sep 17, 2018, 02:04

added 30m of time spent at 2018-09-17

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Sep 17, 2018, 07:48

Hi [Tim] ( @tim.traylor ),  

We have found a new scenario during development which can cause issues in future, so this will cost us additional development time. `1.5 hr`

**Scenario**: If an invoice will be deleted/reverted then `last_assessed_for_service_date` should also to be reverted to the last one.

This will add 2 more test cases on the list: 

* Delete/Revert invoices should also revert the `last_assessed_for_service_date` for that recurring fee.  
* Delete/Revert billing cycle should also revert the `last_assessed_for_service_date` for that recurring fee.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Sep 17, 2018, 07:48

changed time estimate to 1h 30m

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Sep 18, 2018, 07:48

* [x]  Check if the recurring fee is repeating itself in same billing cycle by creating two manual invoices.
* [x]   Check if the recurring fee is repeating itself in same billing cycle by creating one invoice from usage process and one manual.
* [x]  Test quarterly recurring fee in the Monthly billing cycle.
* [x]  Test quarterly recurring fee in the Quarterly billing cycle.
* [x]  Test when account turn on date comes into action.
* [x]  Delete/Revert invoices should also revert the `assessed_date` for that recurring fee.
* [x]  Delete/Revert billing cycle should also revert the `assessed_date` for that recurring fee.


**Note**: It may take one billing cycle to fully take effect.




Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Sep 18, 2018, 23:25

Hi [Kyle] ( @kdjstudios ), 

We request you to please check it ASAP as we switch daily on the specific branch for the particular testing.

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Sep 19, 2018, 05:40

added 5h of time spent at 2018-09-19

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Sep 19, 2018, 05:40

changed time estimate to 4h

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Sep 19, 2018, 08:32

@pchaudhary 

My testing:

1. Logged into MERGE2
2. Clicked billing cycles and clicked "create new" for the Demo Site.
3. Opened Kyle 3 account and added a staged fee "Test Fee 1". Clicked Save.
4. Clicked "Create Invoice" and the below was generated. NO STAGED FEE WAS ADDED
![image](/uploads/fdb7cd1518beb08152de8210ff95e7f0/image.png)
**Failed testing.**

Decided to continue testing due to the vague note in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1037#note_27028 " **Note**: It may take one billing cycle to fully take effect."
 - Please clarify this? If these means that this and/or new issue like highlighted above is still going to happen then this is not a fix we can release. 

5. Clicked back into the account's staged fees and the fee had been removed.
6. Added new staged fee and clicked "save"
7. Clicked back into the staged fees and the fee is not saved. **FAIL**

@tim.traylor @pchaudhary @amishra @smasih 

Team, how was this tested? If I can not even save a staged fee, there is no way your team tested this at all!?

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Sep 20, 2018, 05:29

Hi [Kyle] ( @kdjstudios ),  

Below are the answers to your all concerns:  


> - Logged into MERGE2
>
> - Clicked billing cycles and clicked "create new" for the Demo Site.
>
> - Opened Kyle 3 account and added a staged fee "Test Fee 1". Clicked Save.
>
> - Clicked "Create Invoice" and the below was generated. NO STAGED FEE WAS ADDED

* We would like to tell you that on Step 3, the staged fees were not saved on your case because there was a validation issue while you tested that.  

We also would like to inform you that there is nothing to do with the staged fees as we have already addressed the staged fee concern on #1066.


> Decided to continue testing due to the vague note in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1037#note_27028 " **Note**: It may take one billing cycle to fully take effect."

* *We're unable to get your point by saying vague note*:  As we have already clarified that the fix may take one billing cycle to take full effect because we're checking the `last_assessed_fee_date` for the recurring fees & it's get update when it's applied first.


> Team, how was this tested? If I can not even save a staged fee, there is no way your team tested this at all!?

* We would like to inform you that the issue was fixed for the recurring fees & the staged fees is not the recurring one. So, we request you to test with the recurring fees of the account.

Hope this helps.


Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Sep 20, 2018, 08:04

@pchaudhary 

Okay, so while testing this we found out that adding a staged fee is incurring a 'validation' error that is not being displayed to the user. Please elaborate.

Also, in regards to having to test this on a quarterly account. Which account did you test this on and how? I only show a yearly billing period and don't show a year both of new billing cycles being processed to test this. Please remember nothing should be sent to me until your team has fully tested it. 

![image](/uploads/061b68120c555b9e782cab11d8ffad9a/image.png)

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Sep 21, 2018, 24:47

Hi [Kyle] ( @kdjstudios ),

> Okay, so while testing this we found out that adding a staged fee is incurring a 'validation' error that is not being displayed to the user. Please elaborate.

* Yes, this is a bug. But as we previously mentioned that we have nothing to do with the Staged fees for this issue due to which we did not encounter any such issue while development. Although, we can address it further if you want us to do in this issue otherwise we can create a separate ticket for this? 


> Also, in regards to having to test this on a quarterly account. Which account did you test this on and how? 

* We have tested this on multiple accounts, including the Demo's and El Paso's accounts as well. We have tested the points mentioned in our [comment] ( http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1037#note_27028) & in order to test those points we need to make the *monthly/quarterly* billing cycle & recurring fees both.  


> I only show a yearly billing period and don't show a year both of new billing cycles being processed to test this. Please remember nothing should be sent to me until your team has fully tested it.

* We would like to mention that we delete/restore database multiple times in a day for the development purpose, so there's always a chance that you do not receive the same DB on which we test our cases.  

So, we request you to create a *Monthly/Quarterly* **billing cycle** and **recurring fees** both (if there'll be not) & then test the cases by creating a manual invoice as well as by upload a usage file.

* Usage file for Demo Site. [vcc_test_export.csv](/uploads/1f49765069504ed1a5d56eed61653ffa/vcc_test_export.csv.20180918092200.20180918092200.csv)

* In order to check the last 2 points of our [comment] (http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1037#note_27028), you have to delete the manual invoice on the 1^st case and revert the billing cycle on 2^nd case & then check that creating again the manual invoice or by uploading the usage file should add the recurring fees if the case allows doing the same.

Hope this helps.

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Sep 21, 2018, 10:39

@pchaudhary @abakshi @rchauhan @amishra @tim.traylor @smasih 

Thank you for the update. I have provided a separate ticket for the newly discovered bug. #1135 

In regards to testing, before I am to test anything. Please update this ticket with proof you have fully tested this and it is passing your tests. 

#Key points we should see in you test case results:

- Check if the recurring fee is repeating itself in same billing cycle by creating two manual invoices. **- Simply provide screenshots of the invoices or do not delete the database and provide the invoice numbers so we may check them.**

- Check if the recurring fee is repeating itself in same billing cycle by creating one invoice from usage process and one manual.  **- Simply provide screenshots of the invoices or do not delete the database and provide the invoice numbers so we may check them.** 

- Test quarterly recurring fee in the Monthly billing cycle. **- Simply provide screenshots of the invoices that the quarterly recurring fee is on; or do not delete the database and provide the invoice numbers so we may check them.**

- Test quarterly recurring fee in the Quarterly billing cycle. **- Simply provide screenshots of the invoices that the quarterly recurring fee is on; or do not delete the database and provide the invoice numbers so we may check them.**

- Test when account turn on date comes into action. **- Simply provide screenshots of an account settings with a turn on date after the billing cycle date and provide screen shots of the invoices; or if do not delete the database and provide the account number and invoice numbers so we may check them.**

- Delete/Revert invoices should also revert the `assessed_date` for that recurring fee. **- This must be tested by the Aavaz team as the assessed date is not view able in the GUI. Please test this and provide the necessary proof of success.**

- Delete/Revert billing cycle should also revert the `assessed_date` for that recurring fee. **- This must be tested by the Aavaz team as the assessed date is not view able in the GUI. Please test this and provide the necessary proof of success.**

Updating priority on this ticket. If you have any questions on this please discuss immediately with @tim.traylor

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Sep 24, 2018, 01:11

moving this to discussion for Tuesday call (_25^th Sep 2018_)

**we really need to discuss the testing process, so that every one is on the same page and whole "approval and testing process" goes smoothly.**

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Sep 24, 2018, 01:14

@pchaudhary can we please know the reason of additional `1.5h`

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Sep 26, 2018, 03:45

@smasih 

There is a back & forth development while we were testing it internally which causes the exceed of 1 hour & the remaining half hour was spent on the internal discussion of the process.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Sep 26, 2018, 10:23

@smasih @pchaudhary @amishra @tim.traylor 

Please be sure to bring this up during discussion in your morning meetings with Tim and I believe we should discuss this on tomorrows meeting too.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 24, 2019, 03:42

@kdjstudios 

I went through the whole tickets & here is my understanding:

* We have developed the required functionality & need to test with the current running code. (git/version management overhead)

* I have to test it internally by merging the old branch code with our new code.  

* & if needed then make some changes.


Estimate: 

|Size|Lower Time Bound|Variability|Upper Time Bound|Testing|Time for Testing| Time for Dev & Testing|
|-|-|-|-|-|-|-|
|Medium-1|4|NA|4|High Impact|3.0|7|

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 24, 2019, 10:53

@pchaudhary 

Please review this with @tim.traylor in your morning meetings.  4 hours to merge code and 3 hours to test seems very high from existing code. Especially since it only took 4 hours to develop the code initially.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Feb 4, 2019, 11:20

@smasih @amishra @pchaudhary 

Please be sure to review this with @tim.traylor further, but we need this fixed. This bug that is causing clients to be incorrectly billed.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on May 22, 2019, 01:57

removed milestone

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on May 23, 2019, 07:15

Hi [Kyle] ( @kdjstudios ), 

*We have merged this code with the latest production code & data. We request you to test this on your end & feel free to provide any feedback.*

* [x]  Check if the recurring fee is repeating itself in same billing cycle by creating two manual invoices.
* [x]   Check if the recurring fee is repeating itself in same billing cycle by creating one invoice from usage process and one manual.
* [x]  Test quarterly recurring fee in the Monthly billing cycle.
* [x]  Test quarterly recurring fee in the Quarterly billing cycle.
* [x]  Test when account turn on date comes into action.
* [x]  Delete/Revert invoices should also revert the `assessed_date` for that recurring fee.
* [x]  Delete/Revert billing cycle should also revert the `assessed_date` for that recurring fee.


**Note**: It may take one billing cycle to fully take effect after the release on the production server.



Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 23, 2019, 14:40

@pchaudhary 

marking directly for ready since it is a bug fix and does not need to be announced.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jun 10, 2019, 05:29

@kdjstudios 

We have successfully released this on both of the production servers & tested as well. We request you to please test on your own too.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 10, 2019, 08:55

I wrote:

> Thank you for reporting this to us. After in-depth research and testing our support team was able to update SA Billing's functionality on quarterly fees to now ensure they are only charged on the first billing cycle if multiple billing cycles occur during the same quarterly month. Please let us know if you have any questions or concerns.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 10, 2019, 09:11

Gaylan wrote:

> I am glad this has been corrected as July I have two billings.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 19, 2019, 15:09

closed
